main/src/main/scala/sbt/Defaults.scala

```diff
@@ -1991,14 +2019,35 @@ object Defaults extends BuildCommon {
           else Map.empty[File, URL]
         },
         fileInputOptions := Seq("-doc-root-content", "-diagrams-dot-path"),
+        scalacOptions := {
+          val compileOptions = scalacOptions.value
+          val sv = scalaVersion.value
+          val config = configuration.value
+          val projectName = name.value
+          if (ScalaArtifacts.isScala3(sv)) {
+            val project = if (config == Compile) projectName else s"$projectName-$config"
+            Seq("-project", project)
+          } else compileOptions
+        },
+        sources := Def.taskDyn {
+          val originalSources = sources.value
+          val sv = scalaVersion.value
+          if (ScalaArtifacts.isScala3(sv) && originalSources.nonEmpty) Def.task {
+            val _ = compile.value
+            val tastyFiles = classDirectory.value.**("*.tasty").get
+            tastyFiles.map(_.getAbsoluteFile)
+          } else Def.task(originalSources)
+        }.value,
```

Comment:
I think sources should mean `*.scala`, `*.java`, etc.
I suggest we introduce a new key called `tastyFiles` or something, and return `Nil` when it's not Scala 3. That way we can avoid the use of dynamic task too.